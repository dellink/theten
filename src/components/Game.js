import React from 'react';
import { AppState, StyleSheet, View } from 'react-native';
import AsyncStorage from '@react-native-community/async-storage';
import { BOARD_PADDING, BOARD_SIZE, CELL_MARGIN, CELL_SIZE, STORAGE_KEY } from '../config';
import { saveState } from '../helpers';
import { CHECK_GAME_OVER, LOAD_GAME, MOVE_CELL, NEW_GAME, initialState, reducer } from '../reducer';
import Board from './Board';
import GameButton from './GameButton';
import Overlay from './Overlay';
import Score from './Score';

export default function Game() {
  const [state, dispatch] = React.useReducer(reducer, initialState);
  const { isGameOver, record, score, tiles } = state;

  const stateRef = React.useRef();
  stateRef.current = state;

  const onPressNewGame = () => dispatch({ type: NEW_GAME });

  const onMoveCell = (direction, tile) => {
    dispatch({ type: MOVE_CELL, payload: { direction, tile } });

    setTimeout(() => {
      dispatch({ type: CHECK_GAME_OVER });
    }, 200);
  };

  React.useEffect(() => {
    (async () => {
      try {
        const data = await AsyncStorage.getItem(STORAGE_KEY);
        if (data) {
          dispatch({ type: LOAD_GAME, payload: JSON.parse(data) });
        } else {
          dispatch({ type: NEW_GAME });
        }
      } catch (e) {
        console.log(e);
        dispatch({ type: NEW_GAME });
      }
    })();
  }, []);

  React.useEffect(() => {
    const handleAppStateChange = appState => {
      if (appState.match(/inactive/)) {
        saveState({
          ...stateRef.current,
          tiles: stateRef.current.tiles.filter(tile => tile.mergedInto === null),
        });
      }
    };

    AppState.addEventListener('change', handleAppStateChange);
    return () => {
      AppState.removeEventListener('change', handleAppStateChange);
    };
  }, []);

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <View style={styles.headerColumn}>
          <GameButton title="New game" onPress={onPressNewGame} />
        </View>
        <View style={styles.headerColumn}>
          <Score currentValue={score} record={record} />
        </View>
      </View>
      <View>
        <Board tiles={tiles} onMoveCell={onMoveCell} />
      </View>
      {isGameOver && (
        <Overlay message="Game over" buttonText="Try again" onPress={onPressNewGame} />
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    backgroundColor: '#B2E5FA',
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    padding: 20,
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 60,
    width: BOARD_PADDING * 2 + CELL_MARGIN * BOARD_SIZE * 2 + CELL_SIZE * BOARD_SIZE,
  },
  headerColumn: {
    flexDirection: 'column',
    justifyContent: 'space-between',
  },
});
