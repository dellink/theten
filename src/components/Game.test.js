import React from 'react';
import { act, render } from '@testing-library/react-native';
import Game from './Game';

jest.spyOn(React, 'useEffect').mockImplementation(() => {});

describe(Game.name, () => {
  it('should render initial state correctly', () => {
    expect(render(<Game />).asJSON()).toMatchSnapshot();
  });
});
