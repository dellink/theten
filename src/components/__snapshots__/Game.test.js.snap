// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render initial state correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#B2E5FA",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 60,
          "width": 318,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "column",
            "justifyContent": "space-between",
          }
        }
      >
        <
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#01A9F2",
              "borderColor": "#0187D0",
              "borderRadius": 4,
              "borderWidth": 1,
              "flex": 1,
              "justifyContent": "center",
              "maxHeight": 80,
              "width": 140,
            }
          }
          underlayColor="#0288D1"
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "white",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            New game
          </Text>
        </>
      </View>
      <View
        style={
          Object {
            "flexDirection": "column",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#03A9F4",
              "borderColor": "#0288D1",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "column",
              "minWidth": 120,
              "padding": 10,
              "paddingBottom": 0,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginBottom": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#B3E5FC",
                  "fontFamily": "Verdana",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "marginRight": 10,
                  "textAlign": "center",
                }
              }
            >
              Score:
            </Text>
            <Text
              style={
                Object {
                  "color": "white",
                  "fontFamily": "Verdana",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "textAlign": "center",
                }
              }
            >
              0
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginBottom": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#B3E5FC",
                  "fontFamily": "Verdana",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "marginRight": 10,
                  "textAlign": "center",
                }
              }
            >
              Record:
            </Text>
            <Text
              style={
                Object {
                  "color": "white",
                  "fontFamily": "Verdana",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "textAlign": "center",
                }
              }
            >
              0
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#03A9F4",
            "borderColor": "#0288D1",
            "borderRadius": 5,
            "borderWidth": 1,
            "padding": 3,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "#ddccbb",
                "borderRadius": 5,
                "height": 68,
                "margin": 5,
                "width": 68,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
